sky/clouds/kubernetes.py

```diff
@@ -358,7 +372,19 @@ def accelerator_in_region_or_zone(self,
     def query_status(cls, name: str, tag_filters: Dict[str, str],
                      region: Optional[str], zone: Optional[str],
                      **kwargs) -> List['status_lib.ClusterStatus']:
-        # TODO(romilb): Implement this. For now, we return UP as the status.
-        #  Assuming single node cluster.
         del tag_filters, region, zone, kwargs  # Unused.
-        return [status_lib.ClusterStatus.UP]
+        namespace = kubernetes_utils.get_current_kube_config_context_namespace()
+
+        # Get all the pods with the label skypilot-cluster: <cluster_name>
+        pods = kubernetes.core_api().list_namespaced_pod(
+            namespace, label_selector=f'skypilot-cluster={name}').items
```

Comment:
We should catch the exceptions here to raise only `exceptions.ClusterStatusFetchingError`, as our callers assumes the exceptions to be `exceptions.ClusterStatusFetchingError` for this `query_status` method.
https://github.com/skypilot-org/skypilot/blob/f5526d43b49f5e8bf184f73c58193d9d203cdc1e/sky/backends/backend_utils.py#L2350-L2365
https://github.com/skypilot-org/skypilot/blob/f5526d43b49f5e8bf184f73c58193d9d203cdc1e/sky/backends/backend_utils.py#L2183-L2206

To reproduce:
1. `sky launch --cpus 2+` (on k8s)
2. `sky local down`
3. `sky status -r` fails and my other clouds' cluster fail to show up.
```
$ sky status -r
Clusters
ValueError: Failed to load Kubernetes configuration. Kubeconfig does not contain any valid context(s).
    If you were running a local Kubernetes cluster, run `sky local up` to start the cluster.
```

Reply:
Another question is: if the user switches their `~/.kube/config`, the existing cluster on the original k8s will be leaked if `sky status -r` is called?

That might be fixed by implementing the `get_current_user_identity` in the future, so that the cluster will be marked with the current kubernetes cluster, and will not be operated when it does not match.

Reply:
Ah that's a great catch - I've fixed `sky status -r` now to raise `ClusterStatusFetchingError` when the kubernetes API call fails:
```
# With a empty kubeconfig after sky local down
(base) ➜  ~ sky status -r
Clusters
W 07-25 18:42:14 backend_utils.py:2441] Failed to refresh status for 1 cluster:
W 07-25 18:42:14 backend_utils.py:2444]   test: Failed to query Kubernetes cluster 'test' status: builtins.ValueError: Failed to load Kubernetes configuration. Kubeconfig does not contain any valid context(s).
W 07-25 18:42:14 backend_utils.py:2444]     If you were running a local Kubernetes cluster, run `sky local up` to start the cluster.
NAME  LAUNCHED     RESOURCES                          STATUS  AUTOSTOP  COMMAND
test  36 secs ago  1x Kubernetes(2CPU--2GB, cpus=2+)  UP      -         sky launch --cpus 2+ -c t...

Managed spot jobs
No in progress jobs. (See: sky spot -h)

# With an valid kubeconfig but bad IP address/cluster not reachable
(base) ➜  ~ sky status -r
Clusters
W 07-25 18:43:59 backend_utils.py:2441] Failed to refresh status for 1 cluster:
W 07-25 18:43:59 backend_utils.py:2444]   test: Failed to query cluster 'test' status. Network error - check if the Kubernetes cluster in context gke_skypilot-375900_us-central1-c_testclus2 is up and accessible.
NAME  LAUNCHED    RESOURCES                          STATUS  AUTOSTOP  COMMAND
test  2 mins ago  1x Kubernetes(2CPU--2GB, cpus=2+)  UP      -         sky launch --cpus 2+ -c t...

Managed spot jobs
No in progress jobs. (See: sky spot -h)
```

> Another question is: if the user switches their ~/.kube/config, the existing cluster on the original k8s will be leaked if sky status -r is called?

This is a good point, currently the cluster will indeed be leaked. I've filed #2300 to track and fix this.